### PR TITLE
Add battery pack percent & voltage

### PR DESCRIPTION
**Pull request by SupaFlav**
Add polling for all battery pack voltage & percentage.  Only tested with AC200M, but I imagine it should behave the same with all, or simply be unable to gather the fields.

Additionally, includes the following changes:
  - re-enable switch controls for persistent_conn by employing mutex to guard polling logic
  - add 5x connect attempts to improve consistency of transient & persistent connections
  - persistent connection aids in safeguarding Bluetti device by restricting access to unauthorized users via Bluetooth app

### CODE REVIEW

File: custom_components/bluetti_bt/coordinator.py [hunk 1]
```diff
@@ -180,7 +184,7 @@ async def _async_update_data(self):
         parsed_data: dict = {}
 
         try:
-            async with async_timeout.timeout(15):
+            async with async_timeout.timeout(45):
```
Comment (SupaFlav) [inline custom_components/bluetti_bt/coordinator.py:187]: Maybe 45s is excessive, but max packs is 6 so just added 30s.  For persistent connection, it's really quick, but not sure how long it takes with transient connection.

**Comment by SupaFlav** [comment]
If only Bluetti could keep the packs at the same level ... at least I can monitor them now!
image

**Comment by SupaFlav** [comment]
Totally unrelated, but the voltage levels seem to be a far better indicator of battery level than the actual percentage value provided.  Not much can be done about that as Bluetti will likely never patch it.

I worked on a bit of refactoring for the send/response logic into it's own function, but I need to get ready to head out for the weekend.  I'll try to get back to it on Sunday|Monday to test the changes before pushing.

**Comment by SupaFlav** [comment]
Finally got a chance to test with off & disconnected packs, seems to work fine and AC200M reports 0 volt/percent.  The status field is never actually provided and looks to be only available with the AC300.  Not a big deal and could be made to map off voltage/percent level of pack to report online/offline (can't know charge/discharge it seems).

**Comment by SupaFlav** [comment]
Added additional changes to allow switch controls over a persistent connection.  Employed usage of mutex over checking polling flag (was causing UI lock-ups when both coordinator polling & switch activated).

File: custom_components/bluetti_bt/coordinator.py [hunk 2]
```diff
@@ -183,43 +186,22 @@ async def _async_update_data(self):
...
-
-                    except ParseError:
-                        self.logger.warning("Got a parse exception...")
-                
-                # pack polling
-                for pack in range (1, self.bluetti_device.pack_num_max + 1):
-                    # Set current pack number
-                    await self.async_send_command(
-                        self.bluetti_device.build_setter_command('pack_num', pack)
+        async with self.polling_lock:
```
Comment (SupaFlav) [inline custom_components/bluetti_bt/coordinator.py:189]: same as prior but now guarded by mutex to also be used by switch.py

File: custom_components/bluetti_bt/coordinator.py [hunk 3]
```diff
@@ -265,7 +269,7 @@ async def async_send_command(self, command: ReadHoldingRegisters) -> bytes:
             self.notify_response = bytearray()
 
             # Make request
-            self.logger.debug("Requesting %s (%s)", command)
+            self.logger.debug("Requesting %s", command)
```
Comment (SupaFlav) [inline custom_components/bluetti_bt/coordinator.py:272]: leftover remnant from debugging removed

File: custom_components/bluetti_bt/switch.py [hunk 4]
```diff
@@ -156,38 +159,32 @@ async def async_turn_off(self, **kwargs):
...
-
-        except TimeoutError:
-            _LOGGER.error("Timed out for device %s", mac_loggable(self._address))
-            return None
-        except BleakError as err:
-            _LOGGER.error("Bleak error: %s", err)
-            return None
-        finally:
-            await client.disconnect()
+        async with self._polling_lock:
```
Comment (SupaFlav) [inline custom_components/bluetti_bt/switch.py:164]: Same mutex and connection as used by coordinator.

File: custom_components/bluetti_bt/sensor.py [hunk 5]
```diff
@@ -144,6 +151,7 @@ def _handle_coordinator_update(self) -> None:
 
         response_data = self.coordinator.data.get(self._response_key)
         if response_data is None:
+            _LOGGER.debug("No data for available for (%s)",self._response_key)
```
Comment (Patrick762) [inline custom_components/bluetti_bt/sensor.py:154]: " " missing

**Comment by Patrick762** [comment]
I'll test the PR on the AC300 if I find some spare time. Using mutex is a good idea (I already forgot about this option)

**Comment by SupaFlav** [comment]
Pretty much have everything as I need it now, but I'd like to look into why the initial connection attempt fails so often.  I need to connect with my phone first, let HA timeout on the poll, then disconnect with phone, and HA will connect on the next poll.  It will eventually connect on its own, but it can take quite a bit of time (immediate to 15+ mins on 15s polling).

Another note, the persistent connection can be a security feature as it prevent others from connecting to your device and toggling settings.  Using the Bluetti app shows that no devices are available while HA is connected (desirable IMO).

**Comment by SupaFlav** [comment]
Think I have a fix for my initial connection by adding a reconnect x5 loop. Working like a champ so far with battery pack, switches, quick reconnects after an HA restart. As mentioned before, nobody can connect to the Bluetti while persistent connection is active. A selling feature imo.

Gonna give a few tests later and I'll push it up.

**Comment by SupaFlav** [comment]
Seems to do the trick, but haven't tested for transient connections.  Assuming it should only improve connection there to, but needs to be tested.  As it is right now, everything is working as I need it.  Thanks a bunch for getting this project together!

**Comment by SupaFlav** [comment]
FYI - found a bug after discharging AC200M (let it turn off) where HA had to be restarted to re-establish the connection.  Assuming the Bleak client needs to be recreated to continue using the connection (vs just during init).  This will be an issue for when a device wakes up on PV charge or manually tuned off/on.  Should be tracked separately as this is getting oos.

**Pull request merged**